standards/application/messaging-api.md

```diff
@@ -0,0 +1,965 @@
+  payload: UInt8Array;
+};
+
+interface ISend {
+  public send(encoder: Encoder, message: Message): RequestID;
+  public cancelRequest(requestId: RequestID): void;
+  public getPendingRequests(): RequestID[];
+};
+```
+
+##### `Encoder`
+The `Encoder` object encapsulates the logic for message handling.
+It MUST ensure that messages are encoded into the appropriate format for transmission.
+Additionally, it MUST provide information regarding the `pubsubTopic` and `contentTopic` to which messages SHOULD be sent.
+The specific types of encoders, their extended functionalities, and the mechanisms for their instantiation are considered out of scope for this document.
+
+##### `Message`
+A `Message` object MUST conform to the [MESSAGE](https://github.com/vacp2p/rfc-index/blob/8ee2a6d6b232838d83374c35e2413f84436ecf64/waku/standards/core/14/message.md) specification.
+- The payload property MUST be provided.
+- All other properties are optional; if omitted, they SHOULD be supplied by the `Encoder`.
+
+##### `RequestID`
+A `RequestID` is a `GUID` string that uniquely identifies a message send operation.
+This `GUID` MAY be used to cancel a scheduled request or to retrieve information about the associated message from the `Message Storage API`.
+
+##### `send`
+An `Encoder` and a `Message` MUST be provided as parameters.
+When invoked, the method schedules the message to be sent via [LIGHTPUSH](../standards/core/lightpush.md) or [RELAY](https://github.com/vacp2p/rfc-index/blob/0277fd0c4dbd907dfb2f0c28b6cde94a335e1fae/waku/standards/core/11/relay.md) based on the node’s configuration.
+The method MUST throw an error if the `pubsubTopic` provided by the `Encoder` is not supported by the underlying node.
+The method MUST return the associated `RequestID` for the scheduled send attempt.
+If the send operation fails, it MAY be retried from the `Message Storage API` using the `RequestID`.
+Additionally, send operation can be monitored by subscribing to `message:sent` or `message:error` events described in `Event Source API` below.
+
+##### `cancelRequest`
+Cancels a scheduled send attempt associated with a given `RequestID`.
+A call to this method MUST be ignored if the request has already been fulfilled (i.e., if it has succeeded, failed, or been canceled previously) or if the provided `RequestID` is invalid.
+
+##### `getPendingRequests`
+Returns an array of unique `RequestID` values corresponding to send operations that are still pending.
+A request is considered pending if it has not yet succeeded,
+and it may be actively retrying or scheduled for sending but has not yet been served.
+No duplicate `RequestID` values MUST appear in the returned array.
+
+#### REST API
+
+##### `send`
+Request:
+```http
+POST /send HTTP/1.1
+Accept: application/json
+Content-Type: application/json
+
+{
+  "timestamp": 0,
+  "ephemeral": false,
+  "meta": "ZXhhbXBsZQ==",
+  "pubsubTopic": "string",
+  "contentTopic": "string",
+  "payload": "ZXhhbXBsZQ=="
+}
+```
+
```

Comment:
I would not define things twice. Only once in `C` and provide a recommendation on how this should translate to a REST API

Reply:
@weboko thoughts on this?

IMO

define a C API:

```
char* waku_send(char* message, int shard, int timeoutMs, WakuCallBack onOkCb, WakuCallBack onErrorCb){}
```

and define a way to deduce the REST endpoint:
1. name function because route, `_` becomes `/`
2. data arguments are properties of the JSON payload
3. bytes payload are encoded in base64
4. callbacks are properties on the response JSON/websocket

Which would be:

``
POST /waku/send

{
	message: WakuMessageJSON,
	shard: number,
}
```

etc